edb/edgeql/parser/grammar/convert.py

```diff
@@ -0,0 +1,418 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2008-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import parsing
+import pathlib
+from typing import Dict, List, Sequence, Tuple, Callable
+
+
+class EBNF_Item:
+    pass
+
+
+class EBNF_Literal(EBNF_Item):
+    def __init__(self, token: str):
+        self.token = token
+
+
+class EBNF_Reference(EBNF_Item):
+    def __init__(self, name: str):
+        self.name = name
+
+
+class EBNF_Single(EBNF_Item):
+    def __init__(self, inner: EBNF_Item):
+        self.inner = inner
+
+
+class EBNF_Optional(EBNF_Single):
+    def __init__(self, inner: EBNF_Item):
+        super().__init__(inner)
+
+
+class EBNF_Multiple(EBNF_Item):
+    def __init__(self, inner: Sequence[EBNF_Item]):
+        self.inner = inner
+
+
+class EBNF_Sequence(EBNF_Multiple):
+    def __init__(self, inner: Sequence[EBNF_Item]):
+        super().__init__(inner)
+
+
+class EBNF_Choice(EBNF_Multiple):
+    def __init__(self, inner: Sequence[EBNF_Item]):
+        super().__init__(inner)
+
+
+class EBNF_Production:
+    def __init__(self, name: str, item: EBNF_Item):
+        self.name = name
+        self.item = item
+
+
+def ebnf_single_or_sequence(items: Sequence[EBNF_Item]) -> EBNF_Item:
+    if len(items) == 1:
+        return items[0]
+    else:
+        return EBNF_Sequence(items)
+
+
+def ebnf_single_or_choice(items: Sequence[EBNF_Item]) -> EBNF_Item:
+    if len(items) == 1:
+        return items[0]
+    else:
+        return EBNF_Choice(items)
+
+
+def expand_iso_ebnf(item: EBNF_Item) -> str:
+    if isinstance(item, EBNF_Literal):
```

Comment:
This (and a bunch of other places) would be a perfect time to try out python's new pattern matching system (`match`)